lib/Auth/Process/LogUser.php

```diff
@@ -0,0 +1,224 @@
+{
+    const IDP_KEY = "saml:sp:IdP"; // the key that points to the entity id in the state
+
+    // the metadata key for the IDP's Namespace code (i.e. short name)
+    const IDP_CODE_KEY = 'IDPNamespace';
+
+    const DYNAMO_ENDPOINT_KEY = 'DynamoEndpoint';
+
+    const DYNAMO_REGION_KEY = 'DynamoRegion';
+
+    const DYNAMO_LOG_TABLE_KEY = 'DynamoLogTable';
+
+
+    // The host of the aws dynamodb
+    private $dynamoEndpoint;
+
+    // The region of the aws dynamodb
+    private $dynamoRegion;
+
+    // The name of the aws dynamodb table that stores the login data
+    private $dynamoLogTable;
+
+    /**
+     * Initialize this filter, parse configuration.
+     *
+     * @param array $config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct($config, $reserved) {
+        parent::__construct($config, $reserved);
+        assert(is_array($config));
+
+        if (! empty($config[self::DYNAMO_ENDPOINT_KEY])) {
+            $this->dynamoEndpoint = $config[self::DYNAMO_ENDPOINT_KEY];
+        }
+
+        if (! empty($config[self::DYNAMO_REGION_KEY])) {
+            $this->dynamoRegion = $config[self::DYNAMO_REGION_KEY];
+        }
+
+        if (! empty($config[self::DYNAMO_LOG_TABLE_KEY])) {
+            $this->dynamoLogTable = $config[self::DYNAMO_LOG_TABLE_KEY];
+        }
+
+    }
+
+    /**
+     * Apply filter to copy attributes.
+     *
+     * @param array &$state  The current state array
+     */
+    public function process(&$state) {
+        if (! $this->configsAreValid()) {
+            return;
+        }
+
+        assert('is_array($state)');
+
+        $idp = $this->getIdp($state);
+
+        // Get the SP's entity id
+        $spEntityId = "SP entity ID not available";
+        if (! empty($state['saml:sp:State']['SPMetadata']['entityid'])) {
+            $spEntityId = $state['saml:sp:State']['SPMetadata']['entityid'];
+        }
+
+        // Get the current datetime
+        $datetime = date("Y-m-d H:i:s");
+
+        $sdk = new Aws\Sdk([
+            'endpoint'   => $this->dynamoEndpoint,
+            'region'   => $this->dynamoRegion,
+            'version'  => 'latest'
+        ]);
+
+        $dynamodb = $sdk->createDynamoDb();
+        $marshaler = new Marshaler();
+
+        $userAttributes = $this->getUserAttributes($state);
+
+        $logJson = '
+            {
+                "ID": "' . uniqid() . '",
+                "IDP": "' . $idp . '",
+                "SP": "' . $spEntityId . '",' .
+                $userAttributes .
+                '"Time": "' . $datetime . '"
+            }';
+
+        $item = $marshaler->marshalJson($logJson);
+
+        $params = [
+            'TableName' => $this->dynamoLogTable,
+            'Item' => $item,
+        ];
+
+        try {
+            $result = $dynamodb->putItem($params);
+        } catch (DynamoDbException $e) {
+            echo "Unable to add item:\n";
```

Comment:
Do we know where `echo` output ends up in this context? I would assume it's simply lost to us, but I don't know that for sure. It may be worth changing these to `SimpleSAML\Logger` calls.

Reply:
Fixed